src/interactive-window/interactiveWindow.ts

```diff
@@ -168,72 +164,78 @@ export class InteractiveWindow implements IInteractiveWindowLoadable {
             : notebookEditorOrTab.notebook.uri;
+
         if (!isInteractiveInputTab(notebookEditorOrTab)) {
             this._notebookEditor = notebookEditorOrTab;
         }
 
+        if (preferredController) {
+            this.currentKernelInfo = {
+                controller: preferredController.controller,
+                metadata: preferredController.connection
+            };
+        }
+
         // Set our owner and first submitter
         if (this._owner) {
             this._submitters.push(this._owner);
         }
-    }
 
-    public initialize() {
-        if (this.initialized) {
-            return;
-        }
-        this.initialized = true;
         window.onDidChangeActiveNotebookEditor((e) => {
             if (e === this.notebookEditor) {
                 this._onDidChangeViewState.fire();
             }
         }, this.internalDisposables);
         workspace.onDidCloseNotebookDocument((notebookDocument) => {
-            if (notebookDocument === this.notebookDocument) {
+            if (notebookDocument.uri.toString() === this.notebookUri.toString()) {
                 this.closedEvent.fire();
             }
         }, this.internalDisposables);
 
-        this.cellMatcher = new CellMatcher(this.configuration.getSettings(this.owningResource));
         if (window.activeNotebookEditor === this.notebookEditor) {
             this._onDidChangeViewState.fire();
         }
 
         this.listenForControllerSelection();
-        if (this.preferredController) {
-            // Also start connecting to our kernel but don't wait for it to finish
-            this.startKernel(this.preferredController.controller, this.preferredController.connection).ignoreErrors();
-        } else {
-            traceWarning('No preferred controller found for Interactive Window');
-            if (this.isWebExtension) {
-                this.insertInfoMessage(DataScience.noKernelsSpecifyRemote()).ignoreErrors();
-            }
-        }
 
-        // Create the code generator right away to start watching the notebook
-        this.codeGeneratorFactory.getOrCreate(this.notebookDocument);
+        this.cellMatcher = new CellMatcher(this.configuration.getSettings(this.owningResource));
+        if (this.notebookDocument) {
+            this.codeGeneratorFactory.getOrCreate(this.notebookDocument);
+        }
     }
 
-    public async restore(preferredController: IVSCodeNotebookController | undefined) {
-        if (preferredController) {
-            this.preferredController = preferredController;
-        }
-        if (!this.notebookEditor) {
-            if (isInteractiveInputTab(this.notebookEditorOrTab)) {
-                const document = await workspace.openNotebookDocument(this.notebookEditorOrTab.input.uri);
-                const editor = await window.showNotebookDocument(document, {
-                    viewColumn: this.notebookEditorOrTab.group.viewColumn
+    public async ensureInitialized() {
```

Comment:
We can probably use some typescript type checking trick here since `notebookDocument` can be `undefined`:

```ts
interface IResolvedInteractiveWindow {
    notebookDocument: NotebookDocument;
}

class InteractiveWindow {
    ensureInitialized(): this is IResolvedInteractiveWindow;
}
```